### ISSUE TITLE
About Memory Issues with Large Dataset

### ISSUE THREAD

**Issue by yuanzhiKe** [jrzaurin/pytorch-widedeep#18]
Hello, Javier.
Thank you very much.
You do a great job. It is so easy to use. I love your work. 

I tried to use pytorch-widedeep to learn my own data, where there are about 30M unique ids.
I noticed that the wide model first builds one-hot-vectors and then inputs them to a dense layer, and found that the one-hot-encoder from sklearn costs a lot of memories.
I think maybe it is equal to input the index of the categories to an Embedding layer, which decreases the memory occupation. However, I am not sure how it will affect the performance.

**Comment by jrzaurin** [comment]
Hi @yuanzhiKe 

**MANY** thanks for your kind comments.

Yes, you are right, one hot encoding takes A LOT of memory, moreover for 30M unique ids! 

Feeding indexes to an Embedding layer is pretty much what the `DeepDense` does. The idea of using the `Wide` component is because it directly represents a Linear model. There are two solutions I see I could implement. 

1. Implement a sparse `Wide` component: this is challenging, because to this date I am not aware of an easy way of indexing/slicing sparse tensor fast in Pytorch. 
2. Add an option to "switch off" the `Wide` part. In this case what you get is pretty much the tabular API in fastai, for example (i.e. categorical embeddings combined with numerical columns)

I will explore both solutions soon ;) 

thanks again for your comment

**Comment by jrzaurin** [comment]
Also @yuanzhiKe 

check the docs here: https://pytorch-widedeep.readthedocs.io/en/latest/preprocessing.html#pytorch_widedeep.preprocessing.WidePreprocessor

you have the `sparse` option, but bear in mind that this will be way slower since, as I mention, there is no fast way to slice sparse tensors, therefore, what I do is simply:

```python
    def __getitem__(self, idx: int):
        # X_wide and X_deep are assumed to be *always* present
        if isinstance(self.X_wide, sparse_matrix):
            X = Bunch(wide=np.array(self.X_wide[idx].todense()).squeeze())
        else:
            X = Bunch(wide=self.X_wide[idx])
```

in the data loader, and `todense()` takes time.  I will change this to be more efficient in future versions. If you have any ideas let me know.

**Comment by yuanzhiKe** [comment]
Hi, @jrzaurin 
I think maybe changing the wide model as follows would reduce the memory cost:

```
def __init__(self, vocab_size: int, pred_dim: int = 1):
        super(Wide, self).__init__()
        # self.wide_linear = nn.Linear(wide_dim, pred_dim)
        self.wide_linear = nn.Embedding(vocab_size + 1, pred_dim, padding_idx=0)

def forward(self, X: Tensor) -> Tensor:  # type: ignore
        # out = self.wide_linear(X.float())
        out = self.wide_linear(X).sum(dim=1)
        return out
```

I have forked your code and I am working on it to see whether it changes the performance.

**Comment by jrzaurin** [comment]
Hi @yuanzhiKe 

I have implemented your solution, unit-tested, updated example notebooks, etc...

I need to test al installations in GPU (soon!) and I will open a PR

if you have some time, have a look to the branch `wide_embedding`: https://github.com/jrzaurin/pytorch-widedeep/tree/wide_embedding/pytorch_widedeep

Thanks again for your suggestion!

**Comment by yuanzhiKe** [comment]
Thank you for accepting my solution.
The code changes in .py file in #20 looks good to me
But I find that the loss values in 05_Regression_with_Images_and_Text.ipynb and 06_WarmUp_Model_Components.ipynb are much lower than those at the master branch. Do you know why?

**Comment by jrzaurin** [comment]
Yes, because in the master branch I was printing the RMSE while now I have changed the code to print MSE, for consistency with the `History` callback and the pytorch loss `F.mse_loss`. I have also updated the document and I will explicitly state it in the relase

**Comment by jrzaurin** [comment]
@yuanzhiKe I have now merged and publish to PyPi

Once again, many thanks
